DDF-2520: Fixes windows path issue in storage provider

rzwiefel opened the pull request:
#### What does this PR do?
Windows does not allow ":" in a file path. We assumed that we could translate the URI in storage provider to a native file path. This is not always the case. We now validate each URI to make sure it conforms to our storage provider spec.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
anyone

#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).

[Core APIs](https://github.com/orgs/codice/teams/core-apis)
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@jlcsmith 
@kcwire

jlcsmith commented on catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java:
```diff
@@ -233,6 +233,12 @@ public DeleteStorageResponse delete(DeleteStorageRequest deleteRequest)
                 Path contentIdDir = getContentItemDir(new URI(deletedContentItem.getUri()));
+
+                if (contentIdDir == null) {
+                    throw new StorageException(
```
We shouldn't throw an exception here as it will skip all the other metacards in the delete request.

jlcsmith replied:
In fact, I would also say that the catch block below (line 257) should not re-throw the exception.

jlcsmith commented:
@rzwiefel - we need to consider cases where the metacard has no associated content in the storage provider (eg, where metacard resource-uri is not content:<id>).  It appears that the StorageProvider code you modified assumes that the metacards in the delete request always have an associated item in the content store which is not correct (I could be mis-reading the code though).

rzwiefel commented:
@lessarderic unfortunately I can't make a positive test for it because it is a windows specific issue. Unless we have a strategy for doing that now.

brendan-hofmann commented on catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java:
```diff
@@ -455,9 +466,17 @@ private Path getTempContentItemDir(String requestId, URI contentUri) {
-        return Paths.get(baseContentDirectory.toAbsolutePath()
-                .toString(), pathParts.toArray(new String[pathParts.size()]));
+        try {
+            return Paths.get(baseContentDirectory.toAbsolutePath()
```
I think we want canonical, rather than absolute, paths here. Absolute paths can still contain symlinks, redirects, and OS expansion characters that may or may not be resolved correctly. [This](http://stackoverflow.com/a/1099370) is a good summary.

rzwiefel replied:
Made it so FileSystemStorageProvider Automatically canonicalizes the path when the `baseContentDirectory` is set so everything in here can just use that path as is and it should be :100: (as opposed to everyone repeating the same work every time)

brendan-hofmann replied:
👍

rzwiefel commented:
@brendan-hofmann @lessarderic @jlcsmith 

Updated PR.
 * Updated ContentItemValidator to now only allow valid Content items based on the URI
 * Behavior is now to skip any content item that we "don't own"

There will still be the one issue of handling characters in the URI that are not valid filesystem path characters. https://codice.atlassian.net/browse/DDF-2650

jlcsmith commented on catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java:
```diff
@@ -484,6 +502,9 @@ private Path getContentItemDir(URI contentUri) {
     private Path getContentFilePath(URI uri) throws StorageException {
         Path contentIdDir = getContentItemDir(uri);
+        if (contentIdDir == null) {
+            throw new StorageException("Could find file path for URI :" + uri.toString());
```
nit: typo "Could find..." should be "Could not find..."

rzwiefel replied:
This was not supposed to be in this PR, removing.

jlcsmith commented:
@rzwiefel - are there tests to cover bulk deletes where one of the content items is invalid?

brendan-hofmann commented on catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/content/data/impl/ContentItemValidator.java:
```diff
@@ -26,35 +24,52 @@
  */
 public class ContentItemValidator {
-
     private static final Pattern QUALIFIER_PATTERN = Pattern.compile("\\w[a-zA-Z0-9_\\-]+");
```
Why validate with regexes instead of letting the URI and UUID constructors figure it out for you? Just get and assert on the URI scheme as well and you're done.

rzwiefel replied:
Because flow control with Exceptions is gross. Also Thats what they were doing before and we might as well actually explicitly define (like a regex) what we are expecting. It is far clearer than creating a URI and doing nothing with it purely for the side effect of the exception for validation.

rzwiefel replied:
Also we allow the UUID's to not contain the dashes (and most cases don't) which is how the actual UUID expects the UUID to come in formatted

rzwiefel replied:
They are not the validation we support, so they will not work in this case.

rzwiefel replied:
One of the purposes for this change is because a URI is **too** broad which caused bugs. So we can't use URI. Also as previously stated the UUID will not work since we support no dashes. We really only support content:uuid#qualifier, so hence the regex to validate.

brendan-hofmann replied:
I had forgotten that we strip dashes from the UUID. (as an aside we only do that due to a solr limitation) With that in mind, I see the need for regexes instead.

I'm not sure that we really note this anywhere, or what the correct place to do so would even be. This validator seems like it could use an explanatory comment. As you noted, more naive interpretations have already caused us problems in the past.

rzwiefel replied:
I can most certainly add a note about it here. Also, I do have explicit tests with comments that show we allow dash and dash-less id's which hopefully will help.

pklinef replied:
@brendan-hofmann Stripping dashes from the UUID predates the use of Solr.  I don't know why they were stripped originally.  Maybe @shaundmorris remembers.

rzwiefel commented:
@jlcsmith We don't actually run validation on the delete content items. Not sure if that was intentional or not. I can ask @kcwire  since he was in here last. Any preferences?

brendan-hofmann commented on catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java:
```diff
@@ -459,9 +464,17 @@ private Path getTempContentItemDir(String requestId, URI contentUri) {
+                    pathParts.toArray(new String[pathParts.size()]));
+        } catch (InvalidPathException e) {
+            LOGGER.debug("Invalid path: [{}/{}]",
+                    baseContentDirectory.toAbsolutePath()
```
I know it's just debug logging but  could still use canonical path here to be consistent. 😄

rzwiefel replied:
It should already be canonical path. I can remove the toabsolute path tho 👍

rzwiefel commented:
@jlcsmith Added validation for delete and test for delete
